turbopack/crates/turbo-tasks-backend/src/backend/mod.rs

```diff
@@ -707,18 +719,22 @@ impl<B: BackingStorage> TurboTasksBackendInner<B> {
                     )))
                 }
             };
-            if self.should_track_dependencies()
-                && let Some(reader) = reader
+            if let Some(mut reader_task) = reader_task
                 && options.tracking.should_track(result.is_err())
                 && (!task.is_immutable() || cfg!(feature = "verify_immutable"))
             {
+                let reader = reader.unwrap();
                 let _ = task.add(CachedDataItem::OutputDependent {
                     task: reader,
                     value: (),
                 });
                 drop(task);
 
-                let mut reader_task = ctx.task(reader, TaskDataCategory::Data);
+                // Note that there is a potential race condition here, if the reader task would be
+                // locked after the task is unlocked. The output might be invalidated at
+                // this point, but it wouldn't find the backward dependency or see an outdated
+                // dependency. This way it would loose an invalidation.
```

Comment:
Not really sure if this is better, but I tried my hand at rewording the comment to be a little clearer.

```suggestion
                // Note: We use `task_pair` earlier to lock the task and its reader at the same
                // time. If we didn't and just locked the reader here, an invalidation could occur
                // between grabbing the locks. If that happened, and if the task is "outdated" or
                // doesn't have the dependency edge yet, the invalidation would be lost.
```